url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/triggers/CronExpression.java

```diff
@@ -0,0 +1,1571 @@
+          throw new ParseException("Increment > 12 : " + incr, i);
+        }
+      } else {
+        incr = 1;
+      }
+
+      addToSet(ALL_SPEC_INT, -1, incr, type);
+      return i;
+    } else if (c == 'L') {
+      i++;
+      if (type == DAY_OF_MONTH) {
+        lastdayOfMonth = true;
+      }
+      if (type == DAY_OF_WEEK) {
+        addToSet(7, 7, 0, type);
+      }
+      if (type == DAY_OF_MONTH && s.length() > i) {
+        c = s.charAt(i);
+        if (c == '-') {
+          ValueSet vs = getValue(0, s, i + 1);
+          lastdayOffset = vs.value;
+          if (lastdayOffset > 30) {
+            throw new ParseException("Offset from last day must be <= 30", i + 1);
+          }
+          i = vs.pos;
+        }
+        if (s.length() > i) {
+          c = s.charAt(i);
+          if (c == 'W') {
+            nearestWeekday = true;
+            i++;
+          }
+        }
+      }
+      return i;
+    } else if (c >= '0' && c <= '9') {
+      int val = Integer.parseInt(String.valueOf(c));
+      i++;
+      if (i >= s.length()) {
+        addToSet(val, -1, -1, type);
+      } else {
+        c = s.charAt(i);
+        if (c >= '0' && c <= '9') {
+          ValueSet vs = getValue(val, s, i);
+          val = vs.value;
+          i = vs.pos;
+        }
+        i = checkNext(i, s, val, type);
+        return i;
+      }
+    } else {
+      throw new ParseException("Unexpected character: " + c, i);
+    }
+
+    return i;
+  }
+
+  protected int checkNext(int pos, String s, int val, int type) throws ParseException {
+
+    int end = -1;
+    int i = pos;
+
+    if (i >= s.length()) {
+      addToSet(val, end, -1, type);
+      return i;
+    }
+
+    char c = s.charAt(pos);
+
+    if (c == 'L') {
+      if (type == DAY_OF_WEEK) {
+        if (val < 1 || val > 7) {
+          throw new ParseException("Day-of-Week values must be between 1 and 7", -1);
+        }
+        lastdayOfWeek = true;
+      } else {
+        throw new ParseException("'L' option is not valid here. (pos=" + i + ")", i);
+      }
+      TreeSet<Integer> set = getSet(type);
+      set.add(val);
+      i++;
+      return i;
+    }
+
+    if (c == 'W') {
+      if (type == DAY_OF_MONTH) {
+        nearestWeekday = true;
+      } else {
+        throw new ParseException("'W' option is not valid here. (pos=" + i + ")", i);
+      }
+      if (val > 31) {
+        throw new ParseException("The 'W' option does not make sense with values larger than 31 (max number of days in a month)", i);
+      }
+      TreeSet<Integer> set = getSet(type);
+      set.add(val);
+      i++;
+      return i;
+    }
+
+    if (c == '#') {
+      if (type != DAY_OF_WEEK) {
+        throw new ParseException("'#' option is not valid here. (pos=" + i + ")", i);
+      }
+      i++;
+      try {
+        nthdayOfWeek = Integer.parseInt(s.substring(i));
+        if (nthdayOfWeek < 1 || nthdayOfWeek > 5) {
+          throw new Exception();
+        }
+      } catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)